components/remote_settings/src/storage.rs

```diff
@@ -148,32 +148,83 @@ impl Storage {
     ) -> Result<()> {
         let tx = self.conn.transaction()?;
 
-        // Delete ALL existing records and metadata for every collection_url
         tx.execute("DELETE FROM records", [])?;
         tx.execute("DELETE FROM collection_metadata", [])?;
+        let max_last_modified = Self::update_record_rows(&tx, collection_url, records)?;
+        Self::update_collection_metadata(&tx, collection_url, max_last_modified)?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Merge new records with records stored in the database
+    ///
+    /// Records with `deleted=false` will be inserted into the DB, replacing any previously stored
+    /// records with the same ID. Records with `deleted=true` will be removed.
+    pub fn merge_records(
+        &mut self,
+        collection_url: &str,
+        records: &Vec<RemoteSettingsRecord>,
+    ) -> Result<()> {
+        let tx = self.conn.transaction()?;
 
+        // Delete ALL existing records and metadata for with different collection_urls.
+        //
+        // This way, if a user (probably QA) switches the remote settings server in the middle of a
+        // browser sessions, we'll delete the stale data from the previous server.
+        tx.execute(
+            "DELETE FROM records where collection_url <> ?",
+            [collection_url],
+        )?;
+        tx.execute(
+            "DELETE FROM collection_metadata where collection_url <> ?",
+            [collection_url],
+        )?;
+        let max_last_modified = Self::update_record_rows(&tx, collection_url, records)?;
+        Self::update_collection_metadata(&tx, collection_url, max_last_modified)?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Insert/remove/update rows in the records table based on a records list
+    ///
+    /// Returns the max last modified record from the list
+    fn update_record_rows(
+        tx: &Transaction<'_>,
+        collection_url: &str,
+        records: &[RemoteSettingsRecord],
+    ) -> Result<u64> {
```

Comment:
It's slightly awkward that this has to be an associated function, rather than a `&self` method.  It currently has to be like this, since there's already a mutable borrow to get the transaction.  I think it would be easier to split up operations into multiple methods if we moved to the `Dao` pattern, which stores the transaction in an internal field.